File: CAPI/cpp/API/src/Communication.cpp
#include "Communication.h"
#include "utils.hpp"
#include "structures.h"
#include <thread>
#include <mutex>
#include <condition_variable>
#include <memory>

#undef GetMessage
#undef SendMessage
#undef PeekMessage

using grpc::ClientContext;

Communication::Communication(std::string sIP, std::string sPort)
{
    std::string aim = sIP + ':' + sPort;
    auto channel = grpc::CreateChannel(aim, grpc::InsecureChannelCredentials());
    THUAI8Stub = protobuf::AvailableService::NewStub(channel);
}

bool Communication::Move(int64_t teamID, int64_t characterID, int32_t moveTimeInMilliseconds, double angle)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }

    protobuf::MoveRes moveResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufMoveMsg(teamID, characterID, moveTimeInMilliseconds, angle);
    auto status = THUAI8Stub->Move(&context, request, &moveResult);
    if (status.ok())
    {
        return moveResult.act_success();
    }
    else
    {
        return false;
    }
}

bool Communication::Send(int32_t playerID, int32_t toPlayerID, int32_t teamID, std::string message, bool binary)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit)
            return false;
        counter++;
    }
    protobuf::BoolRes sendMessageResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufSendMsg(playerID, toPlayerID, teamID, std::move(message), binary);
    auto status = THUAI8Stub->Send(&context, request, &sendMessageResult);
    if (status.ok())
        return sendMessageResult.act_success();

    else
        return false;
}

bool Communication::EndAllAction(int32_t playerID, int32_t teamID)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes endAllActionsResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufIDMsg(playerID, teamID);
    auto status = THUAI8Stub->EndAllAction(&context, request, &endAllActionsResult);
    if (status.ok())
        return endAllActionsResult.act_success();
    else
        return false;
}

bool Communication::Recover(int32_t playerID, int64_t recover, int32_t teamID)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes recoverResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufRecoverMsg(playerID, recover, teamID);
    auto status = THUAI8Stub->Recover(&context, request, &recoverResult);
    if (status.ok())
        return recoverResult.act_success();
    else
        return false;
}

bool Communication::Produce(int64_t playerID, int64_t teamID)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes produceResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufIDMsg(playerID, teamID);
    auto status = THUAI8Stub->Produce(&context, request, &produceResult);
    if (status.ok())
        return produceResult.act_success();
    else
        return false;
}

bool Communication::Construct(int32_t playerID, int32_t teamID, THUAI8::ConstructionType constructionType)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes constructResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufConstructMsg(playerID, teamID, constructionType);
    auto status = THUAI8Stub->Construct(&context, request, &constructResult);
    if (status.ok())
        return constructResult.act_success();
    else
        return false;
}

bool Communication::InstallEquipment(int32_t playerID, int32_t teamID, THUAI8::EquipmentType equipmentType)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes installEquipmentResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufEquipMsg(playerID, teamID, equipmentType);
    auto status = THUAI8Stub->Equip(&context, request, &installEquipmentResult);
    if (status.ok())
        return installEquipmentResult.act_success();
    else
        return false;
}

/* bool Communication::Rebuild(int32_t playerID, int32_t teamID, THUAI8::ConstructionType constructionType)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit || counterMove >= moveLimit)
            return false;
        counter++;
        counterMove++;
    }
    protobuf::BoolRes rebuildResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufRebuildMsg(playerID, teamID, constructionType);
    auto status = THUAI8Stub->Rebuild(&context, request, &rebuildResult);
    if (status.ok())
        return rebuildResult.act_success();
    else
        return false;
}*/

// bool Communication::Common_Attack(int32_t playerID, int32_t teamID, double angle)
// {
//     {
//         std::lock_guard<std::mutex> lock(mtxLimit);
//         if (counter >= limit)
//             return false;
//         counter++;
//     }
//     protobuf::BoolRes commonAttackResult;
//     ClientContext context;
//     auto request = THUAI82Proto::THUAI82ProtobufCommonAttackMsg(playerID, teamID, angle);
//     auto status = THUAI8Stub->CommonAttack(&context, request, &commonAttackResult);
//     if (status.ok())
//         return commonAttackResult.act_success();
//     else
//         return false;
// }
// 普攻必中，angle改toplayerID

bool Communication::Common_Attack(int64_t playerID, int64_t teamID, int64_t attacked_playerID, int64_t attacked_teamID)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit)
            return false;
        counter++;
    }
    protobuf::BoolRes commonAttackResult;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufAttackMsg(playerID, teamID);
    auto status = THUAI8Stub->Attack(&context, request, &commonAttackResult);
    if (status.ok())
        return commonAttackResult.act_success();
    else
        return false;
}

bool Communication::BuildCharacter(int32_t teamID, THUAI8::CharacterType charactertype, int32_t birthIndex)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit)
            return false;
        counter++;
    }
    protobuf::BoolRes reply;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufCreatCharacterMsg(teamID, charactertype, birthIndex);
    auto status = THUAI8Stub->CreatCharacter(&context, request, &reply);
    if (status.ok())
        return reply.act_success();
    else
        return false;
}

/*/ bool Communication::Recycle(int32_t playerID, int32_t teamID)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit)
            return false;
        counter++;
    }
    protobuf::BoolRes reply;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufIDMsg(playerID, teamID);
    auto status = THUAI8Stub->(&context, request, &reply);
    if (status.ok())
        return reply.act_success();
    else
        return false;
}*/

// bool Communication::Skill_Attack(int32_t playerID, int32_t teamID, double angle)
// {
//     {
//         std::lock_guard<std::mutex> lock(mtxLimit);
//         if (counter >= limit)
//             return false;
//         counter++;
//     }
//     protobuf::BoolRes reply;
//     ClientContext context;
//     auto request = THUAI82Proto::THUAI82ProtobufSkillAttackMsg(playerID, teamID, angle);
//     auto status = THUAI8Stub->SkillAttack(&context, request, &reply);
//     if (status.ok())
//         return reply.act_success();
//     else
//         return false;
// }
// 技能必中，angle改toplayerID

// 待修改，不知道要不要toteamID
bool Communication::Skill_Attack(int64_t teamID, int64_t playerID, double angle)
{
    {
        std::lock_guard<std::mutex> lock(mtxLimit);
        if (counter >= limit)
            return false;
        counter++;
    }
    protobuf::BoolRes reply;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufCastMsg(playerID,teamID,angle);
    auto status = THUAI8Stub->Cast(&context, request, &reply);
    if (status.ok())
        return reply.act_success();
    else
        return false;
}

bool Communication::TryConnection(int32_t playerID, int32_t teamID)
{
    protobuf::BoolRes reply;
    ClientContext context;
    auto request = THUAI8Proto::THUAI82ProtobufIDMsg(playerID, teamID);
    auto status = THUAI8Stub->TryConnection(&context, request, &reply);
    if (status.ok())
        return true;
    else
        return false;
}

void Communication::AddPlayer(int32_t playerID, int32_t teamID, THUAI8::CharacterType charactertype)
{
    auto tMessage = [=]()
    {
        protobuf::CharacterMsg playerMsg = THUAI8Proto::THUAI82ProtobufCharacterMsg(playerID, teamID, charactertype);
        grpc::ClientContext context;
        auto MessageReader = THUAI8Stub->AddCharacter(&context, playerMsg);

        protobuf::MessageToClient buffer2Client;
        counter = 0;
        counterMove = 0;

        while (MessageReader->Read(&buffer2Client))
        {
            {
                std::lock_guard<std::mutex> lock(mtxMessage);
                message2Client = std::move(buffer2Client);
                haveNewMessage = true;
                {
                    std::lock_guard<std::mutex> lock(mtxLimit);
                    counter++;
                    counterMove = 0;
                }
            }
            cvMessage.notify_one();
        }
    };
    std::thread(tMessage).detach();
}

protobuf::MessageToClient Communication::GetMessage2Client()
{
    std::unique_lock<std::mutex> lock(mtxMessage);
    cvMessage.wait(lock, [this]()
                   { return haveNewMessage; });
    haveNewMessage = false;
    return message2Client;
}

File: CAPI/python/PyAPI/Communication.py
import threading
import time
from typing import Union

import grpc
import proto.Message2Clients_pb2 as Message2Clients
import proto.Services_pb2_grpc as Services
import PyAPI.structures as THUAI8
from PyAPI.AI import Setting
from PyAPI.Interface import IErrorHandler
from PyAPI.utils import THUAI82Proto


class BoolErrorHandler(IErrorHandler):
    @staticmethod
    def result():
        return False


class Communication:
    def __init__(self, sIP: str, sPort: str):
        aim = sIP + ":" + sPort
        channel = grpc.insecure_channel(aim)
        self.__THUAI8Stub = Services.AvailableServiceStub(channel)
        self.__haveNewMessage = False
        self.__cvMessage = threading.Condition()
        self.__message2Client: Message2Clients.MessageToClient
        self.__mtxLimit = threading.Lock()
        self.__counter = 0
        self.__counterMove = 0
        self.__limit = 50
        self.__moveLimit = 10

    def Move(
        self, teamID: int, playerID: int, timeInMiliseconds: int, angle: float
    ) -> bool:
        try:
            with self.__mtxLimit:
                if (
                    self.__counter >= self.__limit
                    or self.__counterMove >= self.__moveLimit
                ):
                    return False
                self.__counter += 1
                self.__counterMove += 1
            moveResult: Message2Clients.MoveRes = self.__THUAI8Stub.Move(
                THUAI82Proto.THUAI82ProtobufMoveMsg(
                    teamID, playerID, timeInMiliseconds, angle
                )
            )
        except grpc.RpcError:
            return False
        else:
            return moveResult.act_success

    def Send(
        self, toID: int, message: Union[str, bytes], playerID: int, teamID: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            sendResult: Message2Clients.BoolRes = self.__THUAI8Stub.Send(
                THUAI82Proto.THUAI82ProtobufSendMsg(
                    playerID,
                    toID,
                    teamID,
                    message,
                    True if isinstance(message, bytes) else False,
                )
            )
        except grpc.RpcError:
            return False
        else:
            return sendResult.act_success

    def Common_Attack(
        self, playerID: int, teamID: int, attackedPlayerID: int, attackedTeamID: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            commonAttackResult: Message2Clients.BoolRes = self.__THUAI8Stub.Attack(
                THUAI82Proto.THUAI82ProtobufAttackMsg(
                    playerID, teamID, attackedPlayerID, attackedTeamID
                )
            )
        except grpc.RpcError:
            return False
        else:
            return commonAttackResult.act_success()

    def Skill_Attack(self, playerID: int, teamID: int, angle: float) -> bool:
        self.__skillrange = 0  # 技能范围待修改,技能位置待定
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            skillAttackResult: Message2Clients.BoolRes = self.__THUAI8Stub.Cast(
                THUAI82Proto.THUAI82ProtobufCastMsg(
                    playerID, 0, teamID, self.__skillrange, 0, 0, angle
                )
            )
        except grpc.RpcError:
            return False
        else:
            return skillAttackResult.act_success()

    def Recover(self, playerID: int, teamID: int, recover: int) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            recoverResult: Message2Clients.BoolRes = self.__THUAI8Stub.Recover(
                THUAI82Proto.THUAI82ProtobufRecoverMsg(playerID, teamID, recover)
            )
        except grpc.RpcError:
            return False
        else:
            return recoverResult.act_success

    def Produce(self, playerID: int, teamID: int) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            produceResult: Message2Clients.BoolRes = self.__THUAI8Stub.Produce(
                THUAI82Proto.THUAI82ProtobufIDMsg(playerID, teamID)
            )
        except grpc.RpcError:
            return False
        else:
            return produceResult.act_success

    def Rebuild(
        self, constructionType: THUAI8.ConstructionType, playerID: int, teamID: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            rebuildResult: Message2Clients.BoolRes = self.__THUAI8Stub.Rebuild(
                THUAI82Proto.THUAI82ProtobufConstructMsg(
                    playerID, teamID, constructionType
                )
            )
        except grpc.RpcError:
            return False
        else:
            return rebuildResult.act_success

    def Construct(
        self, constructionType: THUAI8.ConstructionType, playerID: int, teamID: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            constructResult: Message2Clients.BoolRes = self.__THUAI8Stub.Construct(
                THUAI82Proto.THUAI82ProtobufConstructMsg(
                    playerID, teamID, constructionType
                )
            )
        except grpc.RpcError:
            return False
        else:
            return constructResult.act_success

    def InstallEquipment(
        self, equipmentType: THUAI8.EquipmentType, playerID: int, teamID: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            installEquipmentResult: Message2Clients.BoolRes = self.__THUAI8Stub.Equip(
                THUAI82Proto.THUAI82ProtobufEquipMsg(playerID, teamID, equipmentType)
            )
        except grpc.RpcError:
            return False
        else:
            return installEquipmentResult.act_success

    def EndAllAction(self, playerID: int, teamID: int) -> bool:
        try:
            with self.__mtxLimit:
                if (
                    self.__counter >= self.__limit
                    or self.__counterMove >= self.__moveLimit
                ):
                    return False
                self.__counter += 1
                self.__counterMove += 1
            endResult: Message2Clients.BoolRes = self.__THUAI8Stub.EndAllAction(
                THUAI82Proto.THUAI82ProtobufIDMsg(playerID, teamID)
            )
        except grpc.RpcError:
            return False
        else:
            return endResult.act_success

    def Recycle(self, playerID: int, teamID: int) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            recycleResult: Message2Clients.BoolRes = self.__THUAI8Stub.Recycle(
                THUAI82Proto.THUAI82ProtobufIDMsg(playerID, teamID)
            )
        except grpc.RpcError:
            return False
        else:
            return recycleResult.act_success

    def BuildCharacter(
        self, teamID: int, characterType: THUAI8.CharacterType, birthIndex: int
    ) -> bool:
        try:
            with self.__mtxLimit:
                if self.__counter >= self.__limit:
                    return False
                self.__counter += 1
            buildResult: Message2Clients.BoolRes = self.__THUAI8Stub.CreatCharacter(
                THUAI82Proto.THUAI82ProtobufCreatCharacterMsg(
                    teamID, characterType, birthIndex
                )
            )
        except grpc.RpcError:
            return False
        else:
            return buildResult.act_success

    def TryConnection(self, playerID: int, teamID: int) -> bool:
        try:
            tryResult: Message2Clients.BoolRes = self.__THUAI8Stub.TryConnection(
                THUAI82Proto.THUAI82ProtobufIDMsg(playerID, teamID)
            )
        except grpc.RpcError:
            return False
        else:
            return tryResult.act_success

    def GetMessage2Client(self) -> Message2Clients.MessageToClient:
        with self.__cvMessage:
            self.__cvMessage.wait_for(lambda: self.__haveNewMessage)
            self.__haveNewMessage = False
            return self.__message2Client

    def AddPlayer(
        self,
        playerID: int,
        teamID: int,
        characterType: THUAI8.CharacterType,
        side_flag: bool,
    ) -> None:
        def tMessage():
            try:
                if playerID == 0:
                    playerMsg = THUAI82Proto.THUAI82ProtobufCharacterMsg(
                        playerID, teamID, characterType, side_flag
                    )
                    for msg in self.__THUAI8Stub.AddCharacter(playerMsg):
                        with self.__cvMessage:
                            self.__haveNewMessage = True
                            self.__message2Client = msg
                            self.__cvMessage.notify()
                            with self.__mtxLimit:
                                self.__counter = 0
                                self.__counterMove = 0
                elif playerID >= 1 and playerID <= 6:
                    playerMsg = THUAI82Proto.THUAI82ProtobufCharacterMsg(
                        playerID, teamID, characterType, side_flag
                    )
                    for msg in self.__THUAI8Stub.AddCharacter(playerMsg):
                        with self.__cvMessage:
                            self.__haveNewMessage = True
                            self.__message2Client = msg
                            self.__cvMessage.notify()
                            with self.__mtxLimit:
                                self.__counter = 0
                                self.__counterMove = 0
            except grpc.RpcError:
                return

        threading.Thread(target=tMessage).start()
